### PR TITLE
(fixes #245)(fixes #244) HyperionAwsCli fail fast, and more rs copy o…

### LINKED ISSUE #244: Allow MAXERROR for redshift copy

### LINKED ISSUE #245: HyperionAwsCli need to fail fast
HyperionAwsCli should check for validation error before trying to create the new pipeline.

### PR DESCRIPTION
**Pull request by realstraw**
…ption

### CODE REVIEW

File: core/src/main/scala/com/krux/hyperion/activity/RedshiftCopyActivity.scala [hunk 1]
```diff
@@ -35,10 +36,28 @@ case class RedshiftCopyActivity private (
   failureAndRerunMode: Option[FailureAndRerunMode]
 ) extends PipelineActivity {
 
+  private def validate() =
+    // The following when false will trigger aws datapipeline runtime error:
```
Comment (sethyates) [inline core/src/main/scala/com/krux/hyperion/activity/RedshiftCopyActivity.scala]: Grammar

File: core/src/main/scala/com/krux/hyperion/activity/RedshiftCopyActivity.scala [hunk 2]
```diff
@@ -35,10 +36,28 @@ case class RedshiftCopyActivity private (
...
+    assert(
+      input.dataFormat
```
Comment (sethyates) [inline core/src/main/scala/com/krux/hyperion/activity/RedshiftCopyActivity.scala]: I think you could use `forall` to combine the `map` and `getOrElse` (or maybe `exists`, depending on which way).

File: core/src/main/scala/com/krux/hyperion/activity/RedshiftCopyActivity.scala [hunk 3]
```diff
@@ -35,10 +36,28 @@ case class RedshiftCopyActivity private (
...
+      ,
+      "CSV or TSV format cannot be used with commandOptions"
+    )
+
   def named(name: String) = this.copy(id = id.named(name))
   def groupedBy(group: String) = this.copy(id = id.groupedBy(group))
 
-  def withCommandOptions(opts: RedshiftCopyOption*) = this.copy(commandOptions = commandOptions ++ opts)
+  def withCommandOptions(opts: RedshiftCopyOption*) = {
+    validate()
```
Comment (sethyates) [inline core/src/main/scala/com/krux/hyperion/activity/RedshiftCopyActivity.scala]: This `validate` seems only related to this.  If so, move the `assert` here.

File: core/src/main/scala/com/krux/hyperion/activity/RedshiftCopyOption.scala [hunk 4]
```diff
@@ -22,4 +30,8 @@ object RedshiftCopyOption {
     def repr = Seq("NULL", s"'$nullStr'")
   }
 
+  def maxErrorAs(errorCount: Int) = new RedshiftCopyOption {
```
Comment (sethyates) [inline core/src/main/scala/com/krux/hyperion/activity/RedshiftCopyOption.scala]: `withMaxError`

File: core/src/main/scala/com/krux/hyperion/common/S3Uri.scala [hunk 5]
```diff
@@ -10,7 +10,9 @@ package com.krux.hyperion.common
 case class S3Uri(ref: String) {
   require(ref.startsWith("s3"), "S3Uri must start with s3 protocol.")
 
-  def /(next: String) = S3Uri(s"$ref/$next")
+  def /(next: String): S3Uri = S3Uri(s"$ref/$next")
+
+  def / : S3Uri = /("")
```
Comment (sethyates) [inline core/src/main/scala/com/krux/hyperion/common/S3Uri.scala]: Cool!

File: core/src/main/scala/com/krux/hyperion/datanode/S3DataNode.scala [hunk 6]
```diff
@@ -10,6 +10,8 @@ import com.krux.hyperion.resource.WorkerGroup
 
 sealed trait S3DataNode extends Copyable {
 
+  def dataFormat: Option[DataFormat]
```
Comment (sethyates) [inline core/src/main/scala/com/krux/hyperion/datanode/S3DataNode.scala]: Where is this used/

File: core/src/main/scala/com/krux/hyperion/activity/RedshiftCopyOption.scala (hunk 4 above)
Comment (realstraw) [inline core/src/main/scala/com/krux/hyperion/activity/RedshiftCopyOption.scala]: We are not using `with*` in these ones, and we are using it differently like the following. How about just `maxError`?

```scala
val copyAct = RedshiftCopyActivity(...)(ec2)
      .withCommandOptions(
        RedshiftCopyOption.maxError(maxError),
        RedshiftCopyOption.nullAs("null"),
        RedshiftCopyOption.delimiter(","),
        RedshiftCopyOption.csv
      )
```

File: core/src/main/scala/com/krux/hyperion/datanode/S3DataNode.scala (hunk 6 above)
Comment (realstraw) [inline core/src/main/scala/com/krux/hyperion/datanode/S3DataNode.scala]: the assertion in RedshiftCopyActivity.scala

File: core/src/main/scala/com/krux/hyperion/activity/RedshiftCopyOption.scala (hunk 4 above)
Comment (sethyates) [inline core/src/main/scala/com/krux/hyperion/activity/RedshiftCopyOption.scala]: OK

File: core/src/main/scala/com/krux/hyperion/datanode/S3DataNode.scala (hunk 6 above)
Comment (sethyates) [inline core/src/main/scala/com/krux/hyperion/datanode/S3DataNode.scala]: OK

**Comment by sethyates** [comment]
:sparkles: :santa: :gift: :+1: LGTM :sheep: it

**Pull request closed without merging**